torchdata/datapipes/iter/transform/bucketbatcher.py

```diff
@@ -92,7 +104,116 @@ def __new__(
+    return True
+
+
+@functional_datapipe("max_token_bucketize")
+class MaxTokenBucketizerIterDataPipe(IterDataPipe[DataChunk[T_co]]):
+    r"""
+    Iterable DataPipe to create mini-batches of data from bucket, and the
+    total length of each batch will be limited by ``max_token_count``.
+    (functional name: ``max_token_bucketize``).
+
+    Args:
+        datapipe: Iterable DataPipe being batched
+        max_token_count: Maximum length of total length of data in each batch
+        len_fn: Function to be applied to each element to get lengths. ``len(data)`` is used by default.
+        min_len: Optional minimum length to be included into each batch
+        max_len: Optional maximum length to be included into each batch
+        buffer_size: This restricts how many tokens are taken from prior DataPipe to bucketize
+    """
+    datapipe: IterDataPipe[T_co]
+    max_token_count: int
+    len_fn: Callable
+    min_len: int
+    max_len: Optional[int]
+    buffer_size: int
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe[T_co],
+        max_token_count: int,
+        len_fn: Callable = _default_len_fn,
+        min_len: int = 0,
+        max_len: Optional[int] = None,
+        buffer_size: int = 1000,
+    ) -> None:
+        if min_len < 0:
+            raise ValueError("``min_len`` should be non-negative integer.")
+        if max_len is not None:
+            if min_len > max_len:
+                raise ValueError("``min_len`` should be equal to or smaller than ``max_len``.")
+            if max_len > max_token_count:
+                raise ValueError("``max_token_count`` must be equal to or greater than ``max_len``.")
+        datapipe = datapipe.map(partial(_token_len_fn, len_fn=len_fn))
+        if min_len is not None:
+            datapipe = datapipe.filter(partial(_token_min_filter_fn, min_len=min_len))
+        if max_len is not None:
+            datapipe = datapipe.filter(partial(_token_max_filter_fn, max_len=max_len))
+        if buffer_size <= 0:
+            raise ValueError("'buffer_size' is required to be either None or a positive integer.")
+        self.datapipe = datapipe
+        self.max_token_count = max_token_count
+        self.buffer_size = buffer_size
+
+    def __iter__(self) -> Iterator[DataChunk[T_co]]:
+        buffer: List = []
+        batch: List = []
+        batch_size: int = 0
+        for d in self.datapipe:
+            heapq.heappush(buffer, d)
+            if len(buffer) == self.buffer_size:
+                length, token = heapq.heappop(buffer)
+                if batch_size + length > self.max_token_count:
+                    if not batch:
+                        warnings.warn(
+                            "Empty batch is yielded due to length of a single token is larger than"
+                            "``max_token_count``. Please consider add ``max_len`` to filter such token or"
+                            "increase ``max_token_count``."
+                        )
```

Comment:
This check can be done before the iteration. If ``max_token_count`` is not ``None`` and `max_len`` is not given, we should check the max of the lengths should be shorter than ``max_token_count``.
You can make it by ``datapipe_max_count = datapipe.filter(partial(_token_max_filter_fn, max_len=max_token_count))`` and verify if the length of datapipe is the same as before.

Reply:
Okay, I see. Totally make sense to me.